### PR TITLE
[bitnami/matomo] Add cronjobs for archive and scheduled tasks

### LINKED ISSUE #12580: Matomo cron error
### Name and Version

bitnami/matomo 0.2.5

### What steps will reproduce the bug?

Hello,

I installed the Helm chart and since the container is not running as root, cron does not start.

```
matomo 14:55:31.43
matomo 14:55:31.43 Welcome to the Bitnami matomo container
matomo 14:55:31.43 Subscribe to project updates by watching https://github.com/bitnami/containers
matomo 14:55:31.43 Submit issues and feature requests at https://github.com/bitnami/containers/issues
matomo 14:55:31.43
matomo 14:55:31.44 INFO  ==> ** Starting Matomo setup **
realpath: /bitnami/apache/conf: No such file or directory
matomo 14:55:31.60 INFO  ==> Configuring the HTTP port
matomo 14:55:31.68 INFO  ==> Configuring the HTTPS port
matomo 14:55:31.69 INFO  ==> Configuring Apache ServerTokens directive
matomo 14:55:31.73 INFO  ==> Configuring PHP options
matomo 14:55:31.73 INFO  ==> Setting PHP expose_php option
matomo 14:55:31.87 INFO  ==> Validating settings in MYSQL_CLIENT_* env vars
matomo 14:55:32.09 INFO  ==> Persisted Matomo installation detected
matomo 14:55:32.09 INFO  ==> Updating default plugins in persisted data
matomo 14:55:37.26 INFO  ==> Updating default js scripts in persisted data
matomo 14:55:37.33 INFO  ==> Restoring Matomo installation
matomo 14:55:38.05 INFO  ==> Trying to connect to the database server
matomo 14:55:38.22 INFO  ==> Launching schema update

Everything is already up to date.

Done. To check if this worked please open the system report or run `./console diagnostics:run` and look out for the private directories check. If it doesn't work you may need to execute this command using a user that has write permissions or maybe you are not using Apache or IIS web server. Please note you may need to execut this command every time you update Matomo to a newer version.
matomo 14:55:44.39 WARN  ==> Skipping cron configuration for Matomo because of running as a non-root user
matomo 14:55:44.40 INFO  ==> ** Matomo setup finished! **

matomo 14:55:44.41 WARN  ==> Cron will not be started because of running as a non-root user
matomo 14:55:44.43 INFO  ==> ** Starting Apache **

```

### Are you using any custom parameters or values?

_No response_

### What is the expected behavior?

_No response_

### What do you see instead?

On the UI, shows 1 error:
image

### Additional information

The documentation recommends using cron.
https://matomo.org/faq/on-premise/how-to-set-up-auto-archiving-of-your-reports/

Thank you in advance.

### PR DESCRIPTION
**Pull request by aoterolorenzo**
### Description of the change

In bitnami's chart, matomo's container is run as non-root, what makes it impossible for it's native cron to be executed as described in #12580. This PR adds two different cronjobs for archive and schedule tasks, that would run their cron task with a values-defined schedule, so no root access would be neded (no cron would be executed but the tasks theirselves.

### Applicable issues

- fixes #12580

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

**Comment by artm** [comment]
Large portions of the cronjob specs are copy pasted from the deployment code. Wouldn't it be better if the repeating code were extracted into one or several helper templates reused 3 times (deployment and two cron jobs)? This would make it easier to see the differences between the specs as well as maintain the code.

**Comment by aoterolorenzo** [comment]
> Large portions of the cronjob specs are copy pasted from the deployment code. Wouldn't it be better if the repeating code were extracted into one or several helper templates reused 3 times (deployment and two cron jobs)? This would make it easier to see the differences between the specs as well as maintain the code.

Yes, this is still a draft. I'm firstable flavourizing the cronjobs with bitnami's values, and then will clean and leave the logic to helpers. This is a secondary task on my workload so need to be a little patient here.

**Comment by aoterolorenzo** [comment]
Also I saw that our container logic doesn't follow our usual pattern of executing a different arg than the run.sh after our scripts (if given), which would allow the cronjobs just using our default command with just the arg:

```
              args:
                - /opt/bitnami/php/bin/php /opt/bitnami/matomo/console core:archive
```

Instead we have:

```
# Start Apache
if [[ -f "/opt/bitnami/scripts/nginx-php-fpm/run.sh" ]]; then
    exec "/opt/bitnami/scripts/nginx-php-fpm/run.sh"
else
    exec "/opt/bitnami/scripts/$(web_server_type)/run.sh"
fi
```
So we would need to study the possibility of updating this logic or just manually execute our scripts as in the examples.

Also, we don't currently implement a "wait-for-db" initContainer in the deployment, which would prevent pod restarts till mariadb is up. We will dig in this possibility too.

**Comment by aoterolorenzo** [comment]
> Also I saw that our container logic doesn't follow our usual pattern of executing a different arg than the run.sh after our scripts (if given), which would allow the cronjobs just using our default command with just the arg:
> 
> ```
>               args:
>                 - /opt/bitnami/php/bin/php /opt/bitnami/matomo/console core:archive
> ```
> 
> Instead we have:
> 
> ```
> # Start Apache
> if [[ -f "/opt/bitnami/scripts/nginx-php-fpm/run.sh" ]]; then
>     exec "/opt/bitnami/scripts/nginx-php-fpm/run.sh"
> else
>     exec "/opt/bitnami/scripts/$(web_server_type)/run.sh"
> fi
> ```
> 
> So we would need to study the possibility of updating this logic or just manually execute our scripts as in the examples.
> 

As seen in other assets, is not unexpected to overwrite command and args with full logic (e.g. appsmith initContainers)

> Also, we don't currently implement a "wait-for-db" initContainer in the deployment, which would prevent pod restarts till mariadb is up. We will dig in this possibility too.

The matomo container already has this feature in its bitnami's internal logic, so as we must use this image to trigger the archive and scheduler tasks, a "wait-for-db" container is not needed.

File: bitnami/matomo/templates/_helpers.tpl [hunk 1]
```diff
@@ -142,3 +142,76 @@ mariadb-password
 db-password
 {{- end -}}
 {{- end -}}
+
+{{/*
+Return the matomo image needed initContainers
+*/}}
```
Comment (juan131) [inline bitnami/matomo/templates/_helpers.tpl:148]:
```suggestion
{{/*
Return the matomo pods needed initContainers
*/}}
```

File: bitnami/matomo/templates/_helpers.tpl [hunk 2]
```diff
@@ -142,3 +142,76 @@ mariadb-password
...
+  command:
+    - sh
+    - -c
+    - |
+      mkdir -p "/bitnami/matomo"
+      chown -R "{{ .Values.containerSecurityContext.runAsUser }}:{{ .Values.podSecurityContext.fsGroup }}" "/bitnami/matomo"
+  securityContext:
+    runAsUser: 0
+  {{- if .Values.volumePermissions.resources }}
+  resources: {{- toYaml .Values.volumePermissions.resources | nindent 12 }}
```
Comment (juan131) [inline bitnami/matomo/templates/_helpers.tpl:163]: You're using `nindent 12` here and `nindent 8` when using the helper, see in `deployment.yaml`:
```yaml
        {{ include "matomo.initContainers" . | nindent 8 }}
```
Please be careful and ensure the indentations you use here are relative:
```suggestion
  resources: {{- toYaml .Values.volumePermissions.resources | nindent 4 }}
```

File: bitnami/matomo/templates/_helpers.tpl [hunk 3]
```diff
@@ -142,3 +142,76 @@ mariadb-password
...
+- name: certificates
+  image: {{ template "certificates.image" . }}
+  imagePullPolicy: {{ default .Values.image.pullPolicy .Values.certificates.image.pullPolicy }}
+  imagePullSecrets:
+  {{- range (default .Values.image.pullSecrets .Values.certificates.image.pullSecrets) }}
+    - name: {{ . }}
+  {{- end }}
+  command:
+  {{- if .Values.certificates.command }}
+  command: {{- include "common.tplvalues.render" (dict "value" .Values.certificates.command "context" $) | nindent 12 }}
```
Comment (juan131) [inline bitnami/matomo/templates/_helpers.tpl:179]: Same concern here with `nindent 12`

File: bitnami/matomo/templates/_helpers.tpl [hunk 4]
```diff
@@ -142,3 +142,76 @@ mariadb-password
...
+  - install_packages ca-certificates openssl
+    && openssl req -new -x509 -days 3650 -nodes -sha256
+      -subj "/CN=$(hostname)" -addext "subjectAltName = DNS:$(hostname)"
+      -out  /etc/ssl/certs/ssl-cert-snakeoil.pem
+      -keyout /etc/ssl/private/ssl-cert-snakeoil.key -extensions v3_req
+  {{- end }}
+  {{- if .Values.certificates.args }}
+  args: {{- include "common.tplvalues.render" (dict "value" .Values.certificates.args "context" $) | nindent 12 }}
+  {{- end }}
+  env: {{- include "common.tplvalues.render" (dict "value" .Values.certificates.extraEnvVars "context" $) | nindent 12 }}
```
Comment (juan131) [inline bitnami/matomo/templates/_helpers.tpl:196]: Same concern here with `nindent 12`

File: bitnami/matomo/templates/cronjob.yaml [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+              {{- else }}
+              args:
+                - . /opt/bitnami/scripts/matomo-env.sh &&
+                  . /opt/bitnami/scripts/libbitnami.sh &&
+                  . /opt/bitnami/scripts/liblog.sh &&
+                  /opt/bitnami/scripts/php/setup.sh &&
+                  /opt/bitnami/scripts/mysql-client/setup.sh &&
+                  /opt/bitnami/scripts/matomo/setup.sh &&
+                  /post-init.sh &&
+                  /opt/bitnami/php/bin/php /opt/bitnami/matomo/console core:run-scheduled-tasks
```
Comment (juan131) [inline bitnami/matomo/templates/cronjob.yaml:65]: Feel free to keep it as it is but I'd include new scripts in the container image (e.g. `/opt/bitnami/scripts/matomo/scheduled-tasks.sh` &  `/opt/bitnami/scripts/matomo/archive.sh`) and execute them from here.

File: bitnami/matomo/templates/cronjob.yaml [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+          {{- if .Values.cronjobs.archive.podAnnotations }}
+          annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.cronjobs.archive.podAnnotations "context" $ ) | nindent 12 }}
+          {{- end }}
+        spec:
+          {{- include "matomo.imagePullSecrets" . | nindent 10 }}
+          restartPolicy: OnFailure
+          initContainers:
+            {{ include "matomo.initContainers" . | nindent 12 }}
+          containers:
+            - name: {{ include "common.names.fullname" . }}-scheduled-tasks
```
Comment (juan131) [inline bitnami/matomo/templates/cronjob.yaml:41]: The cronjob name is `name: {{ include "common.names.fullname" . }}-archive` but the container is `{{ include "common.names.fullname" . }}-scheduled-tasks` and runs `core:run-scheduled-tasks` and viceversa in the next cronjob. Please align this properly.

File: bitnami/matomo/templates/cronjob.yaml [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+                {{- if .Values.extraVolumeMounts }}
+                {{- include "common.tplvalues.render" (dict "value" .Values.extraVolumeMounts "context" $) | nindent 12 }}
+                {{- end }}
+              resources:
+                {{- toYaml .Values.resources | nindent 16 }}
+          volumes:
+            {{- if .Values.certificates.customCAs }}
+            - name: etc-ssl-certs
+              emptyDir:
+              medium: "Memory"
```
Comment (juan131) [inline bitnami/matomo/templates/cronjob.yaml:99]: bad indentation?

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for the PR @aoterolorenzo !! Please check my comments when you have a chance.

File: bitnami/matomo/templates/_helpers.tpl (hunk 2 above)
Comment (aoterolorenzo) [reply bitnami/matomo/templates/_helpers.tpl]: Good catch! Fixed [d42183d](https://github.com/bitnami/charts/pull/18754/commits/d42183d1ae6f096356f815db7636069a8e8f068f)

File: bitnami/matomo/templates/_helpers.tpl (hunk 3 above)
Comment (aoterolorenzo) [reply bitnami/matomo/templates/_helpers.tpl]: Fixed [d42183d](https://github.com/bitnami/charts/pull/18754/commits/d42183d1ae6f096356f815db7636069a8e8f068f)

File: bitnami/matomo/templates/cronjob.yaml (hunk 6 above)
Comment (aoterolorenzo) [reply bitnami/matomo/templates/cronjob.yaml]: Fixed [1556933](https://github.com/bitnami/charts/pull/18754/commits/1556933995746dae478e252f55fc5a76378cda98)

File: bitnami/matomo/templates/cronjob.yaml (hunk 7 above)
Comment (aoterolorenzo) [reply bitnami/matomo/templates/cronjob.yaml]: Fixed [6900ca4](https://github.com/bitnami/charts/pull/18754/commits/6900ca41a6dc4bec556ae23ac8cdc759f8c6afc8)

File: bitnami/matomo/templates/cronjob.yaml (hunk 5 above)
Comment (aoterolorenzo) [reply bitnami/matomo/templates/cronjob.yaml]: IMHO I cannot see a huge advantage in case of having the `.sh`'s integrated in the container, since they would become a snippet to maintain with pretty much the includes of the entrypoint, plus the fact once inside the running container, the users could just run both with the single commands:

`/opt/bitnami/php/bin/php /opt/bitnami/matomo/console core:run-scheduled-tasks` 
`/opt/bitnami/php/bin/php /opt/bitnami/matomo/console core:archive` 

Nevertheless, I would indeed see the advantage of having the `run.sh` passthrough as most of the bitnami containers as mentioned before, but I am not sure of the real need of it apart of this case (matomo has a defined passthrough to one or another webserver rn instead the common one)

I would go with leaving this snippets directly on the jobs as in other assets, for now.

File: bitnami/matomo/templates/_helpers.tpl (hunk 1 above)
Comment (aoterolorenzo) [reply bitnami/matomo/templates/_helpers.tpl]: [39dd1ce](https://github.com/bitnami/charts/pull/18754/commits/39dd1ceec22f6133090546d086d59a0d4f8e1957)

**Comment by aoterolorenzo** [comment]
> Thanks for the PR @aoterolorenzo !! Please check my comments when you have a chance.

Thanks for the review! I just updated with some comments accordingly

**Review by juan131 — APPROVED** [review]
LGTM

File: bitnami/matomo/Chart.yaml [hunk 8]
```diff
@@ -38,5 +38,4 @@ maintainers:
   url: https://github.com/bitnami/charts
 name: matomo
 sources:
-- https://github.com/bitnami/charts/tree/main/bitnami/matomo
```
Comment (juan131) [inline bitnami/matomo/Chart.yaml:41]: Why did you remove the sources array?

**Review by juan131 — CHANGES_REQUESTED** [review]

File: bitnami/matomo/Chart.yaml (hunk 8 above)
Comment (aoterolorenzo) [reply bitnami/matomo/Chart.yaml:41]: Seems a bad resolved merge, let's update it. Also, CP has published a fix for the VIB/CI job. Let's merge this!

**Review by juan131 — APPROVED** [review]
LGTM
